peps/pep-0746.rst

```diff
@@ -0,0 +1,149 @@
+PEP: 746
+Title: Type checking Annotated metadata
+Author: Adrian Garcia Badaracco <adrian@adriangb.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Discussions-To: https://discuss.python.org/t/pep-746-typedmetadata-for-type-checking-of-pep-593-annotated/53834
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Created: 20-May-2024
+Python-Version: 3.14
+Post-History: 20-May-2024
+
+Abstract
+========
+
+This PEP proposes a mechanism for type checking metadata that uses
+the :py:data:`typing.Annotated` type. Metadata objects that implement
+the new ``__supports_type__`` protocol will be type checked by static
+type checkers to ensure that the metadata is valid for the given type.
+
+Motivation
+==========
+
+:pep:`593` introduced ``Annotated`` as a way to attach runtime metadata to types.
+In general, the metadata is not meant for static type checkers, but even so,
+it is often useful to be able to check that the metadata makes sense for the given
+type.
+
+Take the first example in :pep:`593`, which uses ``Annotated`` to attach
+serialization information to a field::
+
+   class Student(struct2.Packed):
+       name: Annotated[str, struct2.ctype("<10s")]
+
+Here, the ``struct2.ctype("<10s")`` metadata is meant to be used by a serialization
+library to serialize the field. Such libraries can only serialize a subset of types:
+it would not make sense to write, for example, ``Annotated[list[str], struct2.ctype("<10s")]``.
+Yet the type system provides no way to enforce this. The metadata are completely
+ignored by type checkers.
+
+This use case comes up in libraries like :pypi:`pydantic`, which use
+``Annotated`` to attach validation and conversion information to fields.
+
+Specification
+=============
+
+This PEP introduces a new ``__supports_type__`` protocol that both static and
+runtime type checkers can use to understand if a metadata object in
+``Annotated`` is valid for the given type. Objects that implement this protocol
+must have a method named ``__supports_type__`` that takes a single positional argument and
```

Comment:
"a single positional argument" => If the protocol requires a positional argument, then the parameter should be marked positional-only in the protocol definition. That is, the signature should include a `, /` at the end to indicate that `obj` is positional-only.